### ISSUE TITLE
Missing aggressive app restart capability in background. 

### ISSUE THREAD

**Issue by soloclub** [WhisperSystems/Signal-Android#6456]
I have:
- [x] searched open and closed issues for duplicates
- [x] read https://github.com/WhisperSystems/Signal-Android/wiki/Submitting-useful-bug-reports

----------------------------------------

**Bug description**
Clearing recent apps in my phone kills those apps in recent list. Well it is a bug but some apps like WhatsApp are being restarted aggressively and working while Signal doesn't start and stays inactive.
Aggressive restart behavior of WhatsApp is only observed when phone is connected to Internet.

**Steps to reproduce**
- Open Signal, Whatsapp and settings> app management > running apps.
- Check in running apps both signal and whatsapp are running.
- Minimize all and goto recent app list.
- Clear signal and whatsapp from recent. At this point whatsapp and signal services are killed.
- Quickly switch to settings >app management> running apps

**Actual result:** WhatsApp service shows restarting, Signal is inactive(not listed in running until we open signal app explicitly).
**Expected result:** Signal app service also restarts aggressively in background.

**Screenshots**

![screenshot_2017-03-29-10-31-46](https://cloud.githubusercontent.com/assets/26134720/24439756/d6ea709a-146d-11e7-9951-6f199f71d87d.png)

![screenshot_2017-03-29-10-31-55](https://cloud.githubusercontent.com/assets/26134720/24439769/e8085a68-146d-11e7-80ab-74abe2b4fe68.png)

![screenshot_2017-03-29-10-32-05](https://cloud.githubusercontent.com/assets/26134720/24439777/f1ea5e46-146d-11e7-9f36-c0509904db23.png)

![screenshot_2017-03-29-10-32-37](https://cloud.githubusercontent.com/assets/26134720/24439785/faeebcee-146d-11e7-9a7d-6bbfcd0f8bf6.png)

 **Device info**

**Device:** Lenovo vibe k5 (A6020A40)
**Android version:** 5.1.1 stock
**Signal version:** 4.1.0

**Comment by haffenloher** [comment]
Why would that be necessary? Signal will be restarted as soon as it is needed, i.e. as soon as a push message arrives via GCM.

**Comment by soloclub** [comment]
@haffenloher Well that is not the case here, i.e. Signal service is not being restarted even after sending message and (or) trying a call from other device.

**Comment by soloclub** [comment]
Can anyone confirm looking into the code that Signal app is designed to start running (after being killed) when push notification arrives.

**Comment by moxie0** [comment]
If you force close an app, GCM will not wake it up. Android treats that as the user's intention not to have that app running, I guess.

**Comment by soloclub** [comment]
But what if Signal app is killed by stock background app manager as it is reported in this bug report. I feel signal app should be robust enough to restart in such case similar to Whatsapp or Instagram.

**Comment by moxie0** [comment]
Someone would have to research how that is done and report back here.

**Comment by soloclub** [comment]
I don't know programming but here is something I vaguely understood after searching. Probably START STICKY Service or alarm manager is being used for restarting killed app.

http://developer.android.com/reference/android/app/Service.html#START_STICKY

**Comment by soloclub** [comment]
https://stackoverflow.com/questions/13602014/launch-application-when-gcm-notification-received

**Comment by evilphish** [comment]
I observe the same thing with a couple of friends who like to clear their recent apps list regularly. This leads to them no longer being reachable until they open the signal app manually. I also know of a few people who have ceased using signal exactly because of this behaviour. The reasoning being "With WhatsApp I don't have to remember not to swipe its window away and suddenly end up being unreachable".

This actually does not come as a surprise since if someone installs a messenger he obiously wants to be reachable by his contacts and people expect this behaviour. For all specialized use cases like low battery, do not disturb, etc Android provides ample options and switches with which you can silence an app and thus keep it from notifying or disturbing you. 

At the very least, Signal should include an option in the settings menu to activate aggressive restart behaviour.

**Comment by moxie0** [comment]
I don't know if this will address your underlying issue, but it will do more to keep a service alive.

**Comment by soloclub** [comment]
Signal 4.8.0 beta hasn't fixed this. Issue persists.

**Comment by ibnYusrat** [comment]
This is still a big issue among all of my friends...

**Comment by haffenloher** [comment]
If you're not receiving notifications at all (as opposed to only after force closing), that's a separate issue. Have a look at this support page (especially at no. 9 which is MIUI-specific):

https://support.signal.org/hc/en-us/articles/115005547088-Why-do-I-only-see-new-messages-when-I-open-Signal-Android-

If the issue persists, contact support or try the [forums](https://whispersystems.discoursehosting.net/).
